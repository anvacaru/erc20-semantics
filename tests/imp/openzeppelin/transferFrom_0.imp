//when the recipient is not the zero address
//when the sender has enough approved balance
//when the initial holder has enough balance
//it transfers th requested amount

int a, b, v, caller, initialHolder, initialSupply, spender, amount, to;

initialHolder = 1;
initialSupply = 100;
spender       = 2;
to            = 3;
amount        = initialSupply;

makeThrowAsPrintTrue;
makeAccount(initialHolder, initialSupply);
makeAccount(spender, 0);
makeAccount(to, 0);
makeTotalSupply(initialSupply);

makeAllowance(initialHolder, spender);
test.makeCaller(initialHolder);
test.approve(spender, initialSupply);
test.makeCaller(spender);
test.transferFrom(initialHolder, to, amount);
