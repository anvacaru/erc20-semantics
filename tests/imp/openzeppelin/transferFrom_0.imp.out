Changing the caller to 1
1 is approving allowance 100 for 2
Changing the caller to 2
2 is transferring 100 from 1 to 3
Balance of 1 is 0
Balance of 3 is 100
Allowance of 1 for 2 is 0
<IMP>
  <ERC20>
    <caller>
      2
    </caller>
    <k>
      .
    </k>
    <accounts>
      <account>
        <id>
          1
        </id>
        <balance>
          0
        </balance>
      </account> <account>
        <id>
          2
        </id>
        <balance>
          0
        </balance>
      </account> <account>
        <id>
          3
        </id>
        <balance>
          100
        </balance>
      </account>
    </accounts>
    <allowances>
      <allowance1>
        <owner>
          1
        </owner>
        <spenders>
          <allow>
            <spender>
              2
            </spender>
            <amount>
              0
            </amount>
          </allow>
        </spenders>
      </allowance1> <allowance1>
        <owner>
          2
        </owner>
        <spenders>
          .AllowCellBag
        </spenders>
      </allowance1> <allowance1>
        <owner>
          3
        </owner>
        <spenders>
          .AllowCellBag
        </spenders>
      </allowance1>
    </allowances>
    <log>
      Events: Approval ( 1 , 2 , 100 ) Transfer ( 1 , 3 , 100 )
    </log>
    <supply>
      100
    </supply>
  </ERC20>
  <state>
    a |-> 1
    amount |-> 100
    b |-> 3
    caller |-> 2
    initialHolder |-> 1
    initialSupply |-> 100
    spender |-> 2
    to |-> 3
    v |-> 100
  </state>
  <input>
    ListItem ( #buffer ( "" ) )
    ListItem ( "on" )
    ListItem ( #istream ( 0 ) )
  </input>
  <output>
    ListItem ( #ostream ( 1 ) )
    ListItem ( "on" )
    ListItem ( #buffer ( "" ) )
  </output>
  <throwAsPrint>
    true
  </throwAsPrint>
  <printStrings>
    true
  </printStrings>
</IMP>
